superbench/benchmarks/micro_benchmarks/cublas_function/cublas_benchmark.cu

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "cublas_benchmark.h"
+
+// Cuda context init
+void cuda_init() {
+    CUDA_SAFE_CALL(cudaDeviceReset());
+    CUDA_SAFE_CALL(cudaSetDevice(0));
+    // create streams/handles
+    CUBLAS_SAFE_CALL(cublasCreate(&cublas_handle));
+}
+
+// Cuda context free
+void cuda_free() {
+    CUDA_SAFE_CALL(cudaSetDevice(0));
+    CUBLAS_SAFE_CALL(cublasDestroy(cublas_handle));
+}
+
+template <>
+void gemm<float>(cublasHandle_t handle, int transa, int transb, int m, int n, int k, const float *a, const float *b,
+                 float *c) {
+    float alpha = 1.0f;
+    float beta = 1.0f;
+    CUBLAS_SAFE_CALL(cublasSgemm(handle, (transa ? CUBLAS_OP_T : CUBLAS_OP_N), (transb ? CUBLAS_OP_T : CUBLAS_OP_N), m,
+                                 n, k, &alpha, a, (transa ? k : m), b, (transb ? n : k), &beta, c, m));
+}
+
+template <>
+void gemm<double>(cublasHandle_t handle, int transa, int transb, int m, int n, int k, const double *a, const double *b,
+                  double *c) {
+    double alpha = 1.0;
+    double beta = 1.0;
+    CUBLAS_SAFE_CALL(cublasDgemm(handle, (transa ? CUBLAS_OP_T : CUBLAS_OP_N), (transb ? CUBLAS_OP_T : CUBLAS_OP_N), m,
+                                 n, k, &alpha, a, (transa ? k : m), b, (transb ? n : k), &beta, c, m));
+}
+
+template <>
+void gemm<cuComplex>(cublasHandle_t handle, int transa, int transb, int m, int n, int k, const cuComplex *a,
+                     const cuComplex *b, cuComplex *c) {
+    cuComplex alpha = make_cuComplex(1.0f, 0.0f);
+    cuComplex beta = make_cuComplex(0.0f, 0.0f);
+    CUBLAS_SAFE_CALL(cublasCgemm(handle, (transa ? CUBLAS_OP_T : CUBLAS_OP_N), (transb ? CUBLAS_OP_T : CUBLAS_OP_N), m,
+                                 n, k, &alpha, a, (transa ? k : m), b, (transb ? n : k), &beta, c, m));
+}
+
+template <>
+void gemm<cuDoubleComplex>(cublasHandle_t handle, int transa, int transb, int m, int n, int k, const cuDoubleComplex *a,
```

Comment:
It's very strange that we implement all type of template. Why do we use template?

Reply:
Solved.